librz/arch/isa/msp430/msp430_il.c

```diff
@@ -0,0 +1,199 @@
+            }
+            // the general idea is: First we zero the lower byte through ANDing with 0xFF00
+            //                      Then we assign the lower byte to the (byte-length) result through ORing with 0x00<result>
+            // the overall effect is that only the lower byte is assigned, which is what byte-sized operations do
+            return MSP430_SETR(op->dst, OR(AND(old_value, U16(0xFF00)), UNSIGNED(16, new_value)));
+        } 
+        default: return U16(0); // TODO : other addressing modes
+    }
+}
+
+RzILOpBitVector *update_sr_nz_flags(const RzILOpBitVector *new_value, const RzILOpBitVector *old_sr_value) {
+    // the general idea is that we zero out the N and Z bits in the old SR value 
+    // (by ANDing with a mask of all 1s except in those positions)
+    // and then we OR that resulting value with the new bits in the relevant positions and all 0s everywhere else
+
+    // index of the N flag in the SR register: 2
+    RzILOpBool *n_flag_value = MSB(new_value);
+    RzILOpBitVector *n_or_mask = SHIFTL(IL_FALSE, UNSIGNED(16, n_flag_value), UN8(2));
+    
+    // index of the Z flag in the SR register: 1
+    RzILOpBool *z_flag_value = IS_ZERO(new_value);
+    RzILOpBitVector *z_or_mask = SHIFTL(IL_FALSE, UNSIGNED(16, z_flag_value), UN8(1));
+    
+    RzILOpBitVector *and_mask = UN16(~(1 << 2 | 1 << 1));
+    RzILOpBitVector *or_mask = OR(n_or_mask, z_or_mask);
+    
+    return OR(AND(old_sr_value, and_mask), or_mask);
+}
+
+RzILOpBitVector *update_sr_v_flag_rcc(const RzILOpBitVector *old_value, const RzILOpBool *old_carry, const RzILOpBitVector *old_sr_value) {
+    // the idea is the same as set_nz_flags_from_result(...): we AND with a mask that zeroes out the bit we care about
+    // then we OR with a mask that have the bit we care about in the same position that we zeroed
+    RzILOpBool *v_flag_value = AND(
+        INV(MSB(old_value)),
+        old_carry
+    );
+
+    return OR(
+        AND(old_sr_value, UN16(1 << 8)), // zero out position 8, leave all else the same
+        SHIFTL(IL_FALSE, UNSIGNED(16, v_flag_value), UN8(8)) // OR with the new value for the bit 
+    ); 
+}
+
+RzILOpBitVector *update_sr_c_flag(const RzILOpBool *new_carry, const RzILOpBitVector *old_sr_value) {
+    return MSP430_SETR(MSP430_SR, OR(
+        AND(old_sr_value, UN16(1)), // zero out the 0th position (i.e. the carry bit)
+        UNSIGNED(16, new_carry)    // OR with the new carry 
+    ));
+}
+
+RzILOpEffect *rz_msp430_lift_single_operand_instr(RZ_NONNULL RzAnalysis *analysis, const Msp430Instruction *op) {
+	switch (op->iopcode) {
+    // rotation is just a shift with filling
+    case MSP430_RRC: {
+        // 1- get the carry (to use later as the filling for the MSB of the operand)
+        RzILOpBitVector *sr = MSP430_GETR(MSP430_SR);
+        RzILOpBool *carry = LSB(sr);
+        // 2- get LSB of the operand (register, memory location, ...)
+        RzILOpBitVector *operand = get_destination(op);
```

Comment:
This `operand` variable gets passed to three different functions later. Mind that once a `RzILOpPure` was used in a single operation like `ADD`, you have to use `DUP(operand)` if you want to use it again.
When an rzil operation is freed, it frees all its operands it consumes recursively. So if the same rzil operand (like this one) is used at multiple places without `DUP()` it will be double freed.

Reply:
Adding onto this, in such a case, I would suggest using local variables. Check out the `SETL` and the `VARL` macro. `DUP` literally copies the entire expression, leading to bigger IL. Moreover, it's better if a local variable is used for the same expression rather than duplicating the entire expression, since the former will allow for better analysis.

Reply:
Actually in this case it is even better to use `LET("operand", operand, <expr>)` IMHO.

Reply:
Good Catch! 

Well, I do agree that `DUP`ing is inefficient, but using `LET(...)` is a bit hard because it expects a pure expression as both the value and the body, this turned out to not get along well with what the function is trying to do, which is constructing and returning 2 effects. I ended up using a bunch of `SETL(...)`s in a `SEQ(...)` instead, that was straightforward enough.

Thanks a lot for the free bug-spotting.